### PR TITLE
Feature/issue 3738 norev

### LINKED ISSUE #3738: 2.0 replication doesn't notify clients when unable to send rev message
During pull replication, the following sequence of messages is expected:
1. Sync Gateway sends **changes** message to client.
2. Client returns response indicating which revisions they need.
3. Sync Gateway sends **rev** message for each revision that's requested.
4. Client returns response indicating whether revision was successfully persisted.

In the scenario that Sync Gateway is unable to perform step 3 (most commonly because the revision body is no longer available), it responds by not sending any rev message at all.

Couchbase Lite is currently monitoring replication status by comparing the number of revisions requested in step 2 with the number of rev messages received in step 3.  In the event that Sync Gateway fails to send a rev message, replication becomes stuck in a busy state on the client.  

There are at least two expected scenarios where SG is known to be unable to send a rev, that shouldn't terminate replication:
 - Document has been purged from SG's backing bucket, but is still resident in SG's channel cache
 - Document has been updated on SG under convergence, so old revision body is unavailable 

For expected cases, Sync Gateway could send an rev message with a new `error` property in this scenario to allow CBL to reconcile it's count.  This property could also be used in future to differentiate between error scenarios to trigger additional CBL action.

This doesn't address how to handle unexpected errors (that replication should eventually retry).  I don't know of a mechanism that Lite could do this, other than stopping the entire replication and restarting with the previous since value.  Before triggering that, should Sync Gateway perform a backoff retry on rev message?

### PR DESCRIPTION
**Pull request by tleyden**
Fixes #3738

### CODE REVIEW

**Comment by bbrks** [comment]
Hey @tleyden , if you rebase this on the latest `release/2.1.1` you should get a working branch build (although slow due to current network issues).

File: rest/blip_sync.go [hunk 1]
```diff
@@ -538,14 +538,44 @@ func (bh *blipHandler) handleProposedChanges(rq *blip.Message) error {
...
-		return
+		bh.sendNoRev(err, sender, seq, docID, revID, knownRevs, maxHistory)
+	} else {
+		bh.sendRevision(body, sender, seq, docID, revID, knownRevs, maxHistory)
 	}
+}
+
+func (bh *blipHandler) sendNoRev(err error, sender *blip.Sender, seq db.SequenceID, docID string, revID string, knownRevs map[string]bool, maxHistory int) {
+
+	bh.Logf(base.LevelDebug, base.KeySync, "Sending norev %q %s based on %d known.  User:%s", base.UD(docID), revID, len(knownRevs), base.UD(bh.effectiveUsername))
```
Comment (adamcfraser) [inline rest/blip_sync.go]: I think any logging for the norev message should include the error details.

File: rest/blip_sync.go [hunk 2]
```diff
@@ -538,14 +538,44 @@ func (bh *blipHandler) handleProposedChanges(rq *blip.Message) error {
...
 	if err != nil {
-		base.Warnf(base.KeyAll, "[%s] blipHandler can't get doc %q/%s: %v", bh.blipContext.ID, base.UD(docID), revID, err)
-		return
+		bh.sendNoRev(err, sender, seq, docID, revID, knownRevs, maxHistory)
+	} else {
+		bh.sendRevision(body, sender, seq, docID, revID, knownRevs, maxHistory)
 	}
+}
+
+func (bh *blipHandler) sendNoRev(err error, sender *blip.Sender, seq db.SequenceID, docID string, revID string, knownRevs map[string]bool, maxHistory int) {
```
Comment (adamcfraser) [inline rest/blip_sync.go]: knownRevs and maxHistory aren't relevant for sendNoRev, can be removed.

File: rest/blip_sync.go [hunk 3]
```diff
@@ -570,7 +600,9 @@ func (bh *blipHandler) sendRevision(sender *blip.Sender, seq db.SequenceID, docI
 		outrq.Properties["deleted"] = "1"
 		delete(body, "deleted")
 	}
-	outrq.Properties["sequence"] = string(seqJSON)
+	if marshalErr == nil {
+		outrq.Properties["sequence"] = string(seqJSON)
+	}
```
Comment (adamcfraser) [inline rest/blip_sync.go:605]: Looking at the spec, sequence is generally optional in a rev message, other than the case where a changes message hasn't been sent.  I don't think we have that scenario for rev messages sent by Sync Gateway.   If we can omit sequence altogether, we should file a ticket for that change for Iridium.

**Review by adamcfraser — CHANGES_REQUESTED** [review]

File: rest/blip_sync.go (hunk 1 above)
Comment (tleyden) [reply rest/blip_sync.go]: Fixed

File: rest/blip_sync.go (hunk 2 above)
Comment (tleyden) [reply rest/blip_sync.go]: Fixed

File: rest/blip_sync.go (hunk 3 above)
Comment (adamcfraser) [reply rest/blip_sync.go:605]: Motivation is to reduce CPU cost on SG (marshalling sequence) and transport cost to clients.  Agree that this doesn't need to be changed for this PR.

If there are actually low-level troubleshooting benefits, we should address those via logging.  (although I don't know of any cases where id/rev could be associated with different sequences).

Comment (adamcfraser) [reply rest/blip_sync.go:605]: Looking closer, the sequence that we're sending in the rev message is always just the sequence we previously sent in the changes message - we're not attempting to recalculate it from the document's current metadata.  Given that, I think there's even less value in sending it back to the client on the rev message.  I'm not sure why the changes response from CBL is specifying sequence, in fact.  Should be reviewed w/ https://github.com/couchbase/sync_gateway/issues/3764

**Review by adamcfraser — APPROVED** [review]

**Comment by tleyden** [comment]
The jenkins job finished without issues, so I'm going to go ahead and merge

**Pull request merged**